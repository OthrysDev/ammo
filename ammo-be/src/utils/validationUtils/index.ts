import Sizeof from 'object-sizeof';

const isValidUrl = /((([A-Za-z]{3,9}:(?:\/\/)?)(?:[-;:&=+$,\w]+@)?[A-Za-z0-9.-]+|(?:www\.|[-;:&=+$,\w]+@)[A-Za-z0-9.-]+)((?:\/[+~%/.\w\-_]*)?\??(?:[-+=&;%@.\w_]*)#?(?:[.!/\\\w]*))?)/;

// For Joi to work we must return the object's value, not a boolean
const hasValidWeight = (value: string): string => {
    // Sizeof calculations of bytes based
    const size = Sizeof(value);

    // Arbitrary value of 50ko
    if (size > 50000) {
        throw new Error('your string is too heavy, max 50Ko');
    }

    return value;
};

export { isValidUrl, hasValidWeight };
