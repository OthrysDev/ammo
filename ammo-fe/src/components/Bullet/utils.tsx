import React, { ReactElement } from 'react';
import makeStyles from '@material-ui/core/styles/makeStyles';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';

// ==========================================================
// ==========================================================
// ==========================================================

const useStyles = makeStyles((theme) => ({
    title: {
        color: theme.palette.secondary.light,
        margin: '8px 0',
    },
    values: {
        marginLeft: '24px',
    },
}));

interface TitleAndKeyValuesDescProps {
    id: string;
    title: string;
    obj: Record<string, string>;
}

export const TitleAndKeyValuesDesc = ({
    id,
    title,
    obj,
}: TitleAndKeyValuesDescProps): ReactElement => {
    const classes = useStyles();

    return (
        <Box id={id}>
            <Box id={`${id}-title`} className={classes.title}>
                <Typography variant="subtitle2">{title}</Typography>
            </Box>
            {obj &&
                Object.keys(obj).map((key, i) => {
                    return (
                        <Box
                            id={`${id}-line-${i}`}
                            key={`${id}-value-${i}`}
                            className={classes.values}
                        >
                            <Box>
                                <Typography variant="subtitle2">
                                    <span id={`${id}-line-${i}-key`}>
                                        {key}
                                    </span>{' '}
                                    :{' '}
                                    <span id={`${id}-line-${i}-value`}>
                                        "{obj[key]}"
                                    </span>
                                </Typography>
                            </Box>
                        </Box>
                    );
                })}
        </Box>
    );
};

// ==========================================================

interface TitleAndRawValueDescProps {
    id: string;
    title: string;
    obj: string;
}

export const TitleAndRawValueDesc = ({
    id,
    title,
    obj,
}: TitleAndRawValueDescProps): ReactElement => {
    const classes = useStyles();

    return (
        <Box id={id}>
            <Box id={`${id}-title`} className={classes.title}>
                <Typography variant="subtitle2">{title}</Typography>
            </Box>
            {obj && (
                <Box id={`${id}-value`} className={classes.values}>
                    <Box>
                        <Typography variant="subtitle2">{obj}</Typography>
                    </Box>
                </Box>
            )}
        </Box>
    );
};

// ==========================================================
// ==========================================================
// ==========================================================

const useDelimiterStyles = makeStyles((theme) => ({
    root: {
        display: 'flex',
        width: '100%',
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginTop: '8px',
    },
    border: {
        borderTop: `1px solid ${theme.palette.secondary.dark}`,
        width: '100%',
        marginTop: '10px',
    },
    title: {
        margin: '0 7px',
    },
}));

interface DelimiterProps {
    title: string;
}

export const Delimiter = ({ title }: DelimiterProps): ReactElement => {
    const classes = useDelimiterStyles();

    return (
        <Box className={classes.root}>
            <Box className={classes.border}></Box>
            <Box className={classes.title}>
                <Typography variant="subtitle2">{title}</Typography>
            </Box>
            <Box className={classes.border}></Box>
        </Box>
    );
};
