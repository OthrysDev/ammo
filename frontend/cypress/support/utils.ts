export const isMobile = (): boolean => {
    return (
        Cypress.config('viewportWidth') <
        Cypress.env('mobileViewportWidthBreakpoint')
    );
};

// Alias that is sometimes more suited for use
export const mobile = isMobile;
