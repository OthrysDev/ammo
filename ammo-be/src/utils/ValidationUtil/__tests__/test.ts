import { isValidUrl, hasValidWeight } from 'utils/ValidationUtil';

describe('ValidationUtil', () => {
    describe('URL validation regex', () => {
        test('Feeding a valid URL - should return URL', () => {
            const url = 'http://localhost:3000/user';
            expect(isValidUrl(url)).toBe(url);
        });

        test('Feeding an invalid URL (random string) - should throw', () => {
            expect(() => isValidUrl('AnInvalidUrl')).toThrow();
        });

        test('Feeding an invalid URL (protocol only) - should throw', () => {
            expect(() => isValidUrl('http://')).toThrow();
        });
    });

    describe('Object size validation', () => {
        test('Feeding a valid sized string - should return string', () => {
            expect(hasValidWeight('A not so long string')).toMatchSnapshot();
        });

        test('Feeding an oversized string - should throw', () => {
            expect(() => hasValidWeight('a'.repeat(25001))).toThrow();
        });
    });
});
