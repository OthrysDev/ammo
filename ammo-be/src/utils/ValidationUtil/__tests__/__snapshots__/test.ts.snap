// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationUtil Object size validation Feeding a valid sized string - should return string 1`] = `"A not so long string"`;
